sonolark/lib/kube/kube.go

```diff
@@ -731,3 +741,79 @@ func (m *kubePackage) kubeGet(ctx context.Context, r *apiResource, wait time.Dur
 
 	// not reachable
 }
+
+func (m *kubePackage) kubePortForwardFn(t *starlark.Thread, b *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	if len(args) != 0 {
+		return nil, fmt.Errorf("<%v>: positional args not supported by `kube.portforward': %v", b.Name(), args)
+	}
+
+	if len(kwargs) < 3 {
+		return nil, fmt.Errorf("<%v>: expected at least <resource>=<name> <localport>=<port> <podport>=<port>", b.Name())
+	}
+
+	resource, name, err := getResourceAndName(kwargs[0])
+	if err != nil {
+		return nil, fmt.Errorf("<%v>: %s", b.Name(), err.Error())
+	}
+
+	// If resource is not namespace itself (special case), attempt to parse
+	// namespace out of the arg value.
+	var namespace string
+	if resource != namespaceResrc {
+		ss := strings.Split(name, "/")
+		if len(ss) > 1 {
+			namespace = ss[0]
+			name = ss[1]
+		}
+	}
+
+	var localPort int64
+	var podPort int64
+	for _, kv := range kwargs[1:] {
+		switch string(kv[0].(starlark.String)) {
+		case "localport":
+			var ok bool
+			if localPort, ok = kv[1].(starlark.Int).Int64(); !ok {
+				return nil, fmt.Errorf("<%v>: expected int value for `%s' arg, got: %s", b.Name(), apiGroupKW, kv[1].Type())
+			}
+		case "podport":
+			var ok bool
+			if podPort, ok = kv[1].(starlark.Int).Int64(); !ok {
+				return nil, fmt.Errorf("<%v>: expected int value for `%s' arg, got: %s", b.Name(), apiGroupKW, kv[1].Type())
+			}
+		}
+	}
+
+	roundTripper, upgrader, err := spdy.RoundTripperFor(m.config)
+	if err != nil {
+		return nil, err
+	}
+
+	path := fmt.Sprintf("/api/v1/namespaces/%s/pods/%s/portforward", namespace, name)
+
+	host := strings.Split(m.Master, ":")[0]
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: roundTripper}, http.MethodPost, &url.URL{Scheme: "https", Path: path, Host: host})
+
+	var errorBuffer, outBuffer bytes.Buffer
+	errorWriter := bufio.NewWriter(&errorBuffer)
+	outWriter := bufio.NewWriter(&outBuffer)
+
+	var stopChannel <-chan struct{}
```

Comment:
This is used in the creation, but you haven't provided any means of ending the port forwarding.

To do that you'll have to execute just a few extra pieces of logic:
 - a new kwarg for indicating to stop the forwarding
 - when you create the stopChannel object, you'll want to save it into the thread storage in a way that you can look it up later
 - if they provided the kwarg for stopping it, look up the value and send a message on the stop channel to end it.

Reply:
Since Starlark appears to kill any goroutines after function execution, the user can now specify a path to be tested after starting (and before ending) the port forward for the specified pod. We are simply checking for a 200 response code.